### ISSUE TITLE
Bad performance backfilling large Firestore collections

### ISSUE THREAD

**Issue by davidoort** [algolia/firestore-algolia-search#179]
This snippet of code here:

```typescript
const snapshot = await admin
  .firestore()
  .collectionGroup(collectionPath)
  .offset(offset)
  .limit(DOCS_PER_INDEXING)
  .get();
 ```
 
 Will get executed many times if a collection is large (e.g. 2M records) and it will actually bill offset+DOCS_PER_INDEXING number of reads each time, so after indexing 190k documents, Firestore was showing 80M reads.

Let me know if I'm missing something, but this could be done much more efficiently and cheaply it seems

**Comment by smomin** [comment]
@davidoort if you have a better approach, please go ahead and create a PR and I will review it.

**Comment by go-sean-go** [comment]
Fwiw, the official Firebase docs recommend a cursor strategy: https://firebase.google.com/docs/firestore/query-data/query-cursors - `offset` is not recommended for this sort of case, due to what @davidoort mentioned.

Tl;dr: the simple fix is to grab the last doc in the previous set, and `startAfter()` that doc. (Can also be done w/ values.)

**Comment by davidoort** [comment]
I unfortunately don't have enough bandwidth at the moment to fix this but I would do what @go-sean-go suggests. Do you have time perhaps @go-sean-go ? 🙏

**Comment by go-sean-go** [comment]
Same story here. =/

Our collections are only ~100k docs in size right now, meaning it's probably a few hundred dollars to read them all using the current strategy - assuming @davidoort's math is universal.

At our scale, I've got to just eat the Firebase charge (vs. labor cost to fix it), especially since it only affects us when we sync a new collection on Algolia (which is rare). The main impediments being setting up the project, testing the fix, etc (glancing at the source, the pure code change is probably quick).

@smomin I'm not sure whether you're affiliated w/ Algolia or not (either way, thanks for the lib!), but whether for you or someone else at Algolia, the impact of this is really quite large. To truly understand it, it would be some kind of triangular number calculation (which is beyond my math skills), but I THINK it would be the `triangularNumberFunction(sizeOfCollection / DOCS_PER_INDEXING) * DOCS_PER_INDEXING`. If that math is right (and I believe it is by manually inputting a few smaller numbers), the # of reads required to import collections of various sizes would be...
* 250 docs = 250 reads
* 500 docs = 750 reads
* 750 docs = 1,500 reads
* 1,000 docs = 2,500 reads
* 1,250 docs = 3,750 reads
* 1,500 docs = 5,250 reads
...
* 10,000 docs = 205,000 reads
...
* 100,000 docs = 20,050,000 reads
...
* 1,000,000 docs = 2,000,500,000 reads
* 2,000,000 docs = 8,001,000,000 reads

So, I believe if @davidoort let his collection run fully, it would be about 8 billion reads. At Firestore's current pricing (US multi-region) of $0.06 per 100k reads, that's about $4,800 for one collection import. And 2M docs isn't even that big.

I can't absolutely certify the math above, but I'm pretty sure it's right. Hope that's useful!

**Comment by smomin** [comment]
hey @go-sean-go  I am not against the update but more that my bandwidth is limited.  I am all for improving the extension.  I will see if I can make time but if you do have time to write up a PR, I can get it reviewed and merged.

**Comment by smomin** [comment]
hey @go-sean-go @davidoort I have a PR ready for merging.  Do you guys mind reviewing it?
